readme.md
<img src="img/logo.png" height="15%" width="15%" align="left"/> ***[Storymap.js](https://github.com/jakobzhao/storymap)*** is a javascript library for storytelling with web maps. This library aims to enable individuals, especially those who have little web programming background, to effectively create story map applications over the Internet. Three major tasks would be dealt with, including web interface design, map design and storytelling structure. To develop these functions, this library is written in a combination of several web programming languages, like HTML, Cascading Style Sheets (CSS) and Javascript, and builds upon several prerequisite libraries and web services. To make the story map library robust and lightweight, we select as fewer prerequisite libraries as possible, and prioritize the use of open source or free libraries. This library is currently maintained by [Cartography and Geovisualization Group](http://geoviz.ceoas.oregonstate.edu) at Oregon State University, and has been used in multiple geovisualization related courses (e.g., GEOG 371: Web Mapping, GEOG 571: Advanced Web Mapping, GEOG 472/572: Geovisual Analytics) and projects. In addition, a lot of universities began to choose storymap.js instead of ESRI storymap platform.

A storymap is organically integrated by several scenes. Each scene consists of a web map and a script. You can manipulate the map by zooming, panning, and even adding more thematic layers. This library embodies the concept responsive web design, meaning the storymaps can be shown on either Desktop or mobile devices.

<p>To support open source, buy us a cup of coffee :coffee: :grinning: <a href="https://paypal.me/jakobzhao"><img src="img/Donate-PayPal-green.svg" align="center" /></a> </p>


## Features

A story map consists of several scenes. Each scene has two basic components, in terms of the content object and the map object. The content object is usually made up by texts, graphics, and audios and videos. Here are some key features of this library.

1\. [Scene with fullscreen image or video](https://jakobzhao.github.io/storymap/examples/2.5/video.html)

2\. [MiniGlobe](https://jakobzhao.github.io/storymap/examples/2.5/video.html)

3\. [Nav Bar](https://jakobzhao.github.io/storymap/examples/2.5/navbar.html)

4\. [Map Symbolization](https://jakobzhao.github.io/storymap/examples/2.5/styles.html)

5\. [Tile Layers](https://jakobzhao.github.io/storymap/examples/2.5/tiles.html)

6\. [Data Interation](http://cdn.rawgit.com/jakobzhao/storymap/master/examples/dataInteraction/index.html)

7\. [Spinned loader,progress line, animated scroll-down arrow and more..](hhttps://jakobzhao.github.io/storymap/examples/2.5/video.html)

8\. [3D thematic map](https://cosgrovc.github.io/GEOG4572-PE1v2/index.html);

9\. [Bivariate color pallete generator](http://geoviz.ceoas.oregonstate.edu/storymap/assets/color.html)

## Templates

1\. [Basic Template](https://jakobzhao.github.io/storymap/examples/2.5/helloworld.html);
2\. [Script panel on the right](https://jakobzhao.github.io/storymap/examples/2.5/right.html);
3\. [Split the screen in halves](https://jakobzhao.github.io/storymap/examples/2.5/split.html);
4\. [Atlas Front Page](https://jakobzhao.github.io/storymap/examples/2.5/atlas.html)
5\. [Fading out the scripts not being viewed](https://jakobzhao.github.io/storymap/examples/2.5/fadeout.html)
6\. [arrow lines](https://jakobzhao.github.io/storymap/examples/2.5/swoopy.html)

## Prerequisite

**Required**

1\. [jQuery](https://jquery.com/);

2\. [Bootstrap 4](http://getbootstrap.com/);

3\. [Leaflet](http://leafletjs.com/), (3d Storymap uses [Cesium](http://cesiumjs.org/) instead of leaflet).

4\. [Google Material Icon](https://material.io/icons/)

**Recommended**

1\. Font Awesome;

2\. Animate;

3\. Google Font;

4\. Chroma.js;

5\. leaflet.ajax.min.js.

## Demos

1\. [Oregon Water Atlas](http://oregonwater.info/) -- made by *Garath Baldrica-franklin*

![](img/water.png)

2\. [Stand With #StandingRock](https://winkyt.github.io/standwithstandingrock/) -- made by *Yuanyuan Tian and Yingqi Tang*

![](img/swsr.png)

3\. [Sites in Oregon](https://cdn.rawgit.com/johnsoRiB/storymap_test/1800a77b/PE5/storymap-master/SightsInOregon.html) -- made by *Riley Johnson*

![](img/sites.png)


4\. [Stomping Grounds](http://geoviz.ceoas.oregonstate.edu/mathewsn-storymap/) -- made by *Nicolas Mathews*

![](img/stomping.png)

## Tutorial

A story map is organized as a sequence of scenes, and the scenes are associated with maps. Each map are mashed up by layers, and the layers can be any types of layers that [leaflet.js](http://leafletjs.com/) or cesium support. Below shows the tree structure of this story map library.

```powershell
A Story Map
│
├───Scene 1: Setup
│      │──── Script 1 (e.g., text, video, image, audio, etc.)
│      │──── Map 1
│             │
│             │──── Layer 1
│             │──── Layer 2
│             │──── Layer N
├───Scene 2: Confrontation
│... ... ... ... ... ... ...
│... ... ... ... ... ... ...
│... ... ... ... ... ... ...
├───Scene N: Resolution
│      │──── Content 5
│      │──── Map 5
│             │
│             │──── Layer 1
│             │──── Layer 2
│             │──── Layer N
```

For example, Below is the storytelling structure of a storymap on the standingrock protest at North Dakota, you can see how this story map is organized to embody the setup, the confrontation and then the resolution. The storymap can be found at [here](https://winkyt.github.io/standwithstandingrock/).

![](img/standingrock.png)

Here, we will introduce how to make the [hello world](http://geoviz.ceoas.oregonstate.edu/storymap/examples/2.3/helloworld.html) story map application.

1\. Above all, create an empty html page.

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Hello World!</title>
</head>
<body>

</body>
</html>

```

2\. In the `head` element, please include all the required stylesheets and javascript,the `storymap` stylesheet and library. Please make sure to add stylesheets in front of the javascript libraries.

```html
<head>
    <title>Hello World</title>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!--add required stylesheets-->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css">
    <!--leaflet css-->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.4.0/leaflet.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.5.2/animate.min.css">
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">

    <!--add favicon for the web page-->
    <link rel="shortcut icon" href="../../img/favicon.ico" type="image/x-icon">

    <!--Font-->
    <link href="https://fonts.googleapis.com/css?family=Cairo" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="../../dist/storymap.2.5.css">

    <!--add required libraries-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.4.0/leaflet.js"></script>
    <!--jquery and Popper-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.0/umd/popper.min.js"></script>

    <!--boostrap-->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"></script>
    <!--leaflet.ajax for asynchronously adding geojson data-->
    <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet-ajax/2.1.0/leaflet.ajax.min.js"></script>

    <!--story map plugin-->
    <script src="../../dist/storymap.2.5.js"></script>
</head>
```

To use this storymap library, two files `storymap.2.5.css` and `storymap.2.5.js` are requried to include. Also, this library can be included by download instead of include the source at this repository.

Check [`storymap.2.5.css`](https://raw.githubusercontent.com/jakobzhao/storymap/master/dist/storymap.2.5.css) and [`storymap.2.5.js`](https://raw.githubusercontent.com/jakobzhao/storymap/master/dist/storymap.2.5.js) to download.

3\. Next, in the `body` element, please create a `div` to contain an `storymap` instance.

```html
<body>
<div id="storymap" class="container-fluid" >
    <div class="row">
        <div class="col-sm-6 col-md-8 storymap-map"></div>
        <div class="col-sm-6 col-md-4 storymap-story"></div>
    </div>
</div>
</body>
```

A storymap consists of two key components, namely a map and a story, please put two divs in a `row` div under the container of `storymap`. Please make sure to append class `storymap-map` and `storymap-story` for each div respectively.

4\. Several scenes are contained in a `storymap-story` div. Each scene is held by a `section` element, and assign a `data-scene` attribute. This attribute will be used to link the `section` element to the map component.

```html
<section data-scene="scene1">
    ...
</section>
<section data-scene="scene2">
    ...
</section>
<section data-scene="scene3">
    ...
</section>
```

5\. Each scene holds the content of the scene. various types of multi-media data can be put in the `section` element.

```html
<section data-scene="scene1">
    <h2>scene 1</h2>
</section>
```

6\. For example, if the `data-scene` parameter is `scene1`, the name of the corresponding map object should be scene1` as well. In the layer array object, you will need to include the layer names. For example, layer1, layer2. These layers need to be defined before the scenes object. The `scenes` object can be defined in the script as:

```js
    var scenes = {
        scene1: {lat: 44, lng: -123.5, zoom: 7, layers: [layers.layer2], name: "scene 1"},
        scene2: {lat: 44.5701158, lng: -123.2949388, zoom: 10, layers: [layers.layer2], name: "scene 2"},
        scene3: {lat: 44.5701158, lng: -123.2949388, zoom: 12, layers: [layers.layer1, layers.layer2], name: "scene 3"}
    };
```

Here is the `layers` array:

```js
var layers = {
    layer1: {
        layer: L.tileLayer('http://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}'),
        legend: '<i style="background: black; opacity: 0.5"></i><p><b>legend 1</b></p>'
    },
    layer2: {
        layer: L.geoJson.ajax('http://mapious.ceoas.oregonstate.edu/geoserver/mapious/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=ore_counties&outputFormat=application%2Fjson', {
            color: 'orange',
            weight: 5
        }),
        legend: '<i style="background: orange; opacity: 0.5"></i><p><b>legend 2</b></p>'
    },
    layer3: {
        layer: L.tileLayer('http://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png')
    }
};
```

A storymap has an array of scenes and a scene has an array of layers. The `scene` and `layer` is defined as below.

```js
scene = {
    lat: `float number`,  // the latitude of the center point of the scene.
    lng: `float number`,  // the longitude of the center point of the scene.
    zoom: `integer`,      // the zooming levelr of the map scene.
    layers: `array`,      // the array of layers to add on.
    name: `string`        // the name of the scence This name will be used in the navwidget and/or navbar if any
    script: `function`    // the function to execute at the associated scene.
}
```

```js
layer = {
    layer: `Leaflet Layer`,
    legend: `A legend patch html`
}
```

7\. After defining the layers and scenes, you will create a `storymap` object to anchor to the `storymap` div. We can capture that div using $ selector, Here, the selector is `$('#storymap')`.

```js
$('#storymap').storymap({
    scenes: scenes,
    baselayer: layers.layer3,
    legend: true,
    loader: true,
    flyto: false,
    credits:  "Build with <i class='material-icons' style='color: red; font-size: 10px;'>favorite</i> from John Doe",
    scalebar: true,
    scrolldown: true,
    progressline: true,
    navwidget: true,
    createMap: function () {
        var map = L.map($(".storymap-map")[0], {zoomControl: false}).setView([44, -120], 7);
        basemap = this.baselayer.layer.addTo(map);
        return map;
    }
});
```

>**Note:** To make a storymap, the `scenes` and `layers` objects are required. Other options are optional. By default, the library sets the `breakpoint` **`33%`** from the top of the page, Besides, you can customize your map in the `createMap` function.

Here is the data structure of storymap object.

```js
$(element).storymap({
    triggerpos: `a string of percentage`, // A percentage string'33.333%',
    scenes: `an array of scene objects`,
    baselayer: `a layer`,
    legend: `Boolean value`,
    loader: `Boolean value`,
    flyto: `Boolean value`,
    credits: `html content`,
    scalebar: `Boolean value`,
    scrolldown: `Boolean value`,
    progressline: `Boolean value`,
    navwidget: `Boolean value`,
    createMap: function () {}
});
```

8\. To add a geojson data, you will need to include the `leaflet.ajax.min.js` library in the `head` element. As shown below.

```html
<!--add ajax based data transmission-->
<script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet-ajax/2.1.0/leaflet.ajax.min.js"></script>
```

Till now, in the `script` of the `body` element, you have the following code snippet.

```javascript
<script type="text/javascript">

    var layers = {
        layer1: {
            layer: L.tileLayer('http://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}'),
            legend: '<i style="background: black; opacity: 0.5"></i><p><b>legend 1</b></p>'
        },
        layer2: {
            layer: L.geoJson.ajax('http://mapious.ceoas.oregonstate.edu/geoserver/mapious/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=ore_counties&outputFormat=application%2Fjson', {
                color: 'orange',
                weight: 5
            }),
            legend: '<i style="background: orange; opacity: 0.5"></i><p><b>legend 2</b></p>'
        },
        layer3: {
            layer: L.tileLayer('http://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png')
        }
    };

    var scenes = {
        scene1: {lat: 44, lng: -123.5, zoom: 7, layers: [layers.layer2], name: "scene 1"},
        scene2: {lat: 44.5701158, lng: -123.2949388, zoom: 10, layers: [layers.layer2], name: "scene 2"},
        scene3: {lat: 44.5701158, lng: -123.2949388, zoom: 12, layers: [layers.layer1, layers.layer2], name: "scene 3"}
    };

    $('#storymap').storymap({
        scenes: scenes,
        layers: layers,
        baselayer: layers.layer3,
        legend: true,
        loader: true,
        flyto: false,
        credits: credits:  "Build with <i class='material-icons' style='color: red; font-size: 10px;'>favorite</i> from Bo Zhao",
        scalebar: true,
        scrolldown: false,
        progressline: true,
        navwidget: true,
        createMap: function () {
            var map = L.map($(".storymap-map")[0], {zoomControl: false}).setView([44, -120], 7);
            basemap = this.baselayer.layer.addTo(map);
            return map;
        }
    });
</script>
```

Okay, now we have walk through how to make a simple storymap. See the full source code at [here](https://github.com/jakobzhao/storymap/blob/master/examples/2.5/helloworld.html), and look at this `hello world` application by clicking [here](http://geoviz.ceoas.oregonstate.edu/storymap/examples/2.5/helloworld.html).


## Team

Storymap is currently maintained by the [Cartography and Geovisualization group](geoviz.ceoas.oregonstate.edu) at Oregon State University.

<img src="img/geoviz_osu.png" height="30%" width="30%" align="right"/>

- **[Bo Zhao](https://ceoas.oregonstate.edu/profile/zhao/)** (Project Leader)

- **[Gareth Baldrica-franklin](https://github.com/cartobaldrica)**

## License

This storymap map library is under [the MIT license](LICENSE).


changelog.md
# storymap.js changelog

(all changes without author notice are by [@jakbzhao, Bo Zhao](https://github.com/jakobzhao))

## 2.5 (02/24/2019)

* update the version of the prerequisite javascript libraries.

## 2.4 (05/05/2018)

* add script function for each scene.
* update to Bootstrap 4
* using Material icon instead of glyphicon.


## 2.32 (12/21/2017)

* add a paypal donation badge.

## 2.32 (11/28/2017)

* Change the option name 'breakpointPos to triggerPos'
* change the image out of info panel issue.

## 2.30 (11/04/2017)

* Bivariate Choropleth Pallete Generator.

## 2.30 (10/16/2017)

* Fixed bugs

## 2.25 (05/16/2017)

* Progress Line

## 2.22 (05/14/2017)

* change the navbar class to navwidget, in order to be compatible with Boostraps's navbar elements

## 2.20 (04/14/2017)

* update the scrolling function.

## 2.17.2 (03/27/2017)

* change the place holder.
* enrich the 3d storymap.

## 2.1.11 (03/19/2017)

* make this lib more depends on jquery.
* define the placeholder class.

## 2.1.10 (03/18/2017)

* add a navigation bar.

## 2.1.9 (03/17/2017)

* add animated.css included, then the arrow down icon is animated.
* fullpage feature works on IE.
* update several bugs
* add the home icon for the last scene.

## 2.1.7 (03/17/2017)

* add fullpage feature. only works for chrome.

## 2.1.6 (03/13/2017)

* add .gitignore file to remove .idea folder.

## 2.1.4 (03/01/2017)

* fix the the map view issue when the width is less than 768px
* update the class for button from social to btn
* add the changelog.md

## master

An in-progress version being developed on the `master` branch.


backup/storymap.js
// Modified by Bo Zhao, zhao2@oregonstate.edu
// Originally obtained from http://atlefren.github.io/storymap/
// Updated on 5/14/2017 | version 2.22 | MIT License
(function ($) {

    $.fn.storymap = function(options) {

        var defaults = {
            selector: '[data-scene]',
            breakpointPos: '33.333%',
            legend: false,
            scale: false,
            navwidget: false,
            createMap: function () {
                var map = L.map('map', {zoomControl: false}).setView([44, -120], 7);
                L.tileLayer('http://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png');
                return map;
            }
        };

        var settings = $.extend(defaults, options);

        if (typeof(L) === 'undefined') {
            throw new Error('Storymap requires Laeaflet.');
        }



        function getDistanceToTop(elem, top) {

            var docViewTop = $(window).scrollTop();

            var elemTop = $(elem).offset().top;

            var dist = elemTop - docViewTop;

            var d = top - dist;

            if (d < 0) {
                return $(document).height();
            }
            return d;
        }

        function highlightTopPara(sections, top) {

            var distances = $.map(sections, function (element) {
                var dist = getDistanceToTop(element, top);
                return {el: $(element), distance: dist};
            });

            function findMin(pre, cur) {
                if (pre.distance > cur.distance) {
                    return cur;
                } else {
                    return pre;
                }
            }

            var closest = distances.reduce(findMin);

            $.each(sections, function (key, element) {

                var section = $(element);
                if (section[0] !== closest.el[0]) {
                    section.trigger('notviewing');
                }
            });

            if (!closest.el.hasClass('viewing')) {
                closest.el.trigger('viewing');
            }
        }

        function watchHighlight(element, searchfor, top) {

            var sections = element.find(searchfor);
            highlightTopPara(sections, top);

            $(window).scroll(function () {
                highlightTopPara(sections, top);
            });
        }


        //support video for IE 8 and 9.
        document.createElement('video');

        var makeStoryMap = function (element, scenes, layers) {

            var topElem = $('<div class="breakpoint-current"></div>')
                .css('top', settings.breakpointPos);
            $('body').append(topElem);

            var top = topElem.offset().top - $(window).scrollTop();
            var searchfor = settings.selector;
            var sections = $(element).find(searchfor);


            var map = settings.createMap();
            var currentLayerGroup = L.layerGroup().addTo(map);
            var legendControl = L.control({position: 'topright'}); // you can change the position of the legend Control.

            if (settings.scale) {
                L.control.scale({position: 'bottomright', metric: false}).addTo(map);
            }

            if (!String.prototype.includes) {
                String.prototype.includes = function() {
                    'use strict';
                    return String.prototype.indexOf.apply(this, arguments) !== -1;
                };
            }

            if ($(".navbar").length !== 0) {

                var navbar_height = $(".navbar").height();

                var origin_main_top = $(".main").position().top;

                $(".main").css({ top: (navbar_height + origin_main_top).toString() + "px"});
            }

            $.each(sections, function (key, element) {

                var section = $(element);

                //Update the height of the viewing section by changing the co-efficiency
                if (section.height() <= $(window).height() * 0.33) {
                    section.height($(window).height() * 0.33)
                }

                // if (section[0].className === 'viewing' && scenes[section.data('scene')].position !== "fullpage") {
                //
                //     var scene = scenes[$(section).data('scene')];
                //
                //     map.setView([scene.lat, scene.lng], scene.zoom);
                //
                //     var layernames = scene.layers;
                //     var legendContent = "";
                //
                //     if(typeof layernames !== 'undefined') {
                //
                //         for (var i = 0; i < layernames.length; i++) {
                //
                //             //add new layers
                //             currentLayerGroup.addLayer(layers[layernames[i]][0]);
                //
                //             //add new legends
                //             if (layers[layernames[i]].length === 2) {
                //                 legendContent += layers[layernames[i]][1];
                //             }
                //         }
                //     }
                //
                //     legendControl.onAdd = function () {
                //         var div = new L.DomUtil.create('div', 'legend');
                //         div.innerHTML = legendContent;
                //
                //         return div;
                //     };
                //
                //     if (settings.legend === true && legendContent !== "")
                //     {
                //         legendControl.addTo(map);
                //
                //         if ($(".navbar").length !== 0) {
                //             navbar_height = $(".navbar").height();
                //             origin_legend_top = $(".legend").position().top;
                //             $(".legend").css({ top: (navbar_height + origin_legend_top).toString() + "px"});
                //         }
                //     }
                // }
                //
            } );

            function showMapView(key) {

                currentLayerGroup.clearLayers();

                if (settings.legend === true)
                {
                    legendControl.remove();
                }

                var scene = scenes[key];
                var layernames = scene.layers;
                var legendContent = "";

                if(typeof layernames !== 'undefined' && scene.position !== "fullpage") {
                    for (var i=0; i < layernames.length; i++)
                    {
                        currentLayerGroup.addLayer(layers[layernames[i]][0]);
                        if (layers[layernames[i]].length === 2)  {
                            legendContent += layers[layernames[i]][1];
                        }
                    }
                }

                legendControl.onAdd = function () {
                    var div = new L.DomUtil.create('div', 'legend');
                    div.innerHTML = legendContent;
                    return div;
                };

                // the condition legendContent != "" will make sure the legend will only be added on when there is some contents in the legend.
                if (settings.legend === true && legendContent !== "")
                {
                    legendControl.addTo(map);
                    if ($(".navbar").length !== 0) {
                        navbar_height = $(".navbar").height();
                        origin_legend_top = $(".legend").position().top;
                        $(".legend").css({ top: (navbar_height + origin_legend_top).toString() + "px"});
                    }
                }
                map.setView([scene.lat, scene.lng], scene.zoom, 1);

            }

            sections.on('viewing', function () {


                $(".loader").fadeTo(0,1);

                $(this).addClass('viewing');
                $(".arrow-down").css("left", "2%");

                if (scenes[$(this).data('scene')].position === "fullpage") {
                    $(this).addClass('section-opacity')
                        .css('width', "0px")
                        .css('padding', "0 0 0 0");
                    $(this).find(".background-fullscreen-setting")
                        .addClass('fullpage')
                        .css("display", "block");
                    $(".arrow-down").css("left", "50%");
                }

                // // Change the arrow-down icon to the home icon when reaching the last scene.
                if ($(this).data('scene') === sections.last().data('scene')) {
                    $(".arrow-down").removeClass("glyphicon-menu-down")
                        .addClass("glyphicon-home");

                } else {
                    $(".arrow-down").removeClass("glyphicon-home")
                        .addClass("glyphicon-menu-down");
                }

                // Bounce the arrow-down icon when the icon is on the front page.
                if ($(this).data('scene') === sections.first().data('scene') || $(this).data('scene') === sections.last().data('scene')) {
                    $(".arrow-down").addClass("animated");
                } else {
                    $(".arrow-down").removeClass("animated");
                }

                showMapView($(this).data('scene'));

                $(".loader").fadeTo(500,0);

            });


            sections.on('notviewing', function () {
                $(this).removeClass('viewing');

                if (scenes[$(this).data('scene')].position === "fullpage") {
                    $(this).removeClass('section-opacity');
                    $(this).find(".background-fullscreen-setting")
                        .removeClass('fullpage')
                        .css("display", "none");
                }
            });

            watchHighlight(element, searchfor, top);

            $('.arrow-down').click(function () {
                if ($(".arrow-down")[0].className.includes("menu")) {

                    //if there is a navbar.
                    if ($(".navbar").length !== 0) {

                        window.scrollBy(0, $(".viewing").offset().top -$(window).scrollTop() - $('.navbar').height() + $('.viewing').height() - 10);
                    } else {
                        window.scrollBy(0, $(".viewing").offset().top -$(window).scrollTop() + $('.viewing').height() - 10);
                    }

                } else if ($(".arrow-down")[0].className.includes("home")) {
                    window.scrollTo(0, 0);
                }
            });


            // create a progress line
            $(window).scroll(function(){
                var wintop = $(window).scrollTop(), docheight = $(document).height(), winheight = $(window).height();
                var scrolled = (wintop/(docheight-winheight))*100;

                $('.progress-line').css('width', (scrolled + '%'));
            });

            // create the navigation widget anchored on the side.
            if (settings.navwidget) {
                $.each(sections, function (key, element) {
                    var section = $(element);
                    // if no name attribute for a specific scene, the name on the navigation bar will be the object name.
                    if (typeof(scenes[section.data('scene')].name) === "undefined") {
                        sceneName = section.data('scene');
                    } else {
                        sceneName = scenes[section.data('scene')].name.replace(" ","&nbsp;");
                    }

                    //if there is a navbar.
                    if ($(".navbar").length !== 0) {
                        //scrollScript = "javascript:window.scrollBy(0, $('section[data-scene=\\'" + section.data('scene') + "\\']').offset().top - $(window).scrollTop() - $('.navbar').height() - 10);";
                        scrollScript = "javascript:window.scrollBy(0, $('.viewing').offset().top - $(window).scrollTop() - $('.navbar').height() + $('.viewing').height() - 10);";
                    } else {
                        scrollScript = "javascript:window.scrollBy(0, $('.viewing').offset().top - $(window).scrollTop() + $('viewing').height() - 10);";
                    }

                    if (key == 0) {
                        $(".navwidget").append('<li><a class="glyphicon glyphicon-home" data-toggle="tooltip" style="font-size:16px" title="' + sceneName + '" href="' + scrollScript + '" ></a></li>');
                    } else {
                        $(".navwidget").append('<li><a class="glyphicon glyphicon-one-fine-full-dot" data-toggle="tooltip" title="' + sceneName + '" href="' + scrollScript + '" ></a></li>');
                    }
                });

                $('[data-toggle="tooltip"]').tooltip({placement: 'right', html: true});


                $( ".navwidget" ).hover(function() {
                    $(this).fadeTo( 100, 0.8 );
                }, function() {
                    $(this).fadeTo( 300, 0);
                });
            }

        };

        makeStoryMap(this, settings.scenes, settings.layers);
        return this;
    }

}(jQuery));
